metrics/src/datapoint.rs

```diff
@@ -46,7 +46,7 @@ pub struct DataPoint {
     pub name: &'static str,
     pub timestamp: SystemTime,
     /// tags are eligible for group-by operations.
-    pub tags: Vec<(&'static str, &'static str)>,
+    pub tags: Vec<(&'static str, String)>,
```

Comment:
Given that there is still an additional String copy but we also want to have the flexibility for non-static strings, I am thinking about maybe we have two separate Vec of tags, one for static and one for non-static.  So the syntax can be something like:


"static-tag" => "static-tag-value",
("non-static-tag" => "non-static-tag-value"),

So the take-away is that: for those entries having () is for non-static and could incur copy.  For those without (), it is for static.

How does it sound?

Reply:
> Given that there is still an additional String copy but we also want to have the flexibility for non-static strings, I am thinking about maybe we have two separate Vec of tags, one for static and one for non-static

I agree that a `Vec<&'static str, &'static>` saves a copy for the case where tag-values are static. However, the addition of a second tag `Vec` to `DataPoint` would:
- Add additional memory footprint (extra pointer+length+capacity) of that `Vec` to every `DataPoint`
- Incur runtime cost of checking that second `Vec` in every `DataPoint` when we serialize the datapoint string

Reply:
Hmm, that's also a valid point.

For cases where we want to migrate String fields to string tags, this approach performs better as it offers no additional overhead.  Static string cases should be relatively rare.